### PR TITLE
C# Improving Channel API

### LINKED ISSUE #2508: Implement connectivity API for C#
#2477 introduced the C core API for checking the connectivity state of a channel. This API needs to be wrapped and exposed in any language that exposes a channel-like abstraction.

### LINKED ISSUE #2637: Use NULL for default host in grpc_channel_call_create() in C#
#2635 will introduce a mechanism whereby the default host can be correctly determined by core. Since this can (almost) never be correctly determined automatically by wrapped languages without massive code duplication, wrapped languages are encouraged to use this facility once it's available.

### PR DESCRIPTION
**Pull request by jtattermusch**
-- add channel state API
-- start using default host when calling grpc_channel_create_call
-- add some tests 
-- fix some style nits
-- exposes Target property for channel with the right contents.

Fixes #2508
Fixes #2637

### CODE REVIEW

**Comment by jtattermusch** [comment]
Hm, looks flaky.

**Comment by murgatroid99** [comment]
In tests, for channels where you set the channel argument `grpc.ssl_target_name_override`, make sure you also set `grpc.default_authority`.

File: src/csharp/Grpc.Core.Tests/ChannelTest.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                Assert.Throws(typeof(ArgumentException), () => channel.WaitForStateChangedAsync(ChannelState.FatalFailure));
+            }
+        }
+
+        [Test]
+        public void Target()
+        {
+            using (var channel = new Channel("127.0.0.1", Credentials.Insecure))
+            {
+                Assert.IsTrue(channel.Target.Contains("127.0.0.1"));
```
Comment (murgatroid99) [inline src/csharp/Grpc.Core.Tests/ChannelTest.cs]: The actual value of the channel target is not guaranteed.

Comment (jtattermusch) [inline src/csharp/Grpc.Core.Tests/ChannelTest.cs]: Seems it working reliably on my machine and in our test environment. Do you still suggest removing?

**Comment by jtattermusch** [comment]
pull request #2779 and #2778 should address the flakiness. Rebased this PR to contain them.

**Comment by jtattermusch** [comment]
Why is grpc.default_authority important? Looks like none of gRPC implementations is setting it.

**Comment by jtattermusch** [comment]
MIchael is OOO, I will discuss the grpc.default_authority with him separately. Please review the rest of the code.

**Comment by stanley-cheung** [comment]
Didn't see anything that jumps out to me. LGTM.

**Pull request merged**